docs(idempotency): improve navigation, wording, and new section on guarantees (#4613)
awslabs/aws-lambda-powertools-python: docs/utilities/idempotency.md

<!-- markdownlint-disable MD051 -->

> The idempotency utility allows you to retry operations with the same input within a time window, producing the same output.

Reviewer: At this point, the customer may not be familiar with the concept of idempotency and the idea of "producing the same output" could lead to misunderstandings. It's important to clarify that the same output refers to the first sucessful invocation. 

To avoid confusion, why not use the same explanation as before and just add the time window sentence?

Author: I think this fits betters, but I'll try a few more this week.

```
The idempotency utility allows you to retry operations within a time window with the same input, producing the same output.
```

## Key features

> * Produces the same result when a function is called repeatedly with the same idempotency key

Reviewer: Same as before here.. Same result for a failed execution? And if I update the code to fix the bug and send the same payload?

Author: fixed it

* Choose your idempotency key from one or more fields, or entire payload
* Safeguard concurrent requests, timeouts, missing idempotency keys, and payload tampering

> * Support for Amazon DynamoDB, Redis, and bring your own persistence layer

Reviewer: You explain in the "Terminology" section that "in-memory caching" is supported. Does it make sense to add it here?

Author: good idea. I was trying to not turn into a kitchen sink. Adding this here instead of creating another line 

<img width="840" alt="image" src="https://github.com/user-attachments/assets/d2f974a8-2641-4739-9e66-d67afe920b3c">

## Terminology

The property of idempotency means that an operation does not cause additional side effects if it is called more than once with the same input parameters.

**Idempotency key** is a combination of **(a)** Lambda function name, **(b)** fully qualified name of your function, and **(c)** a hash of the entire payload or part(s) of the payload you specify.

**Idempotent request** is an operation with the same input previously processed that is not expired in your persistent storage or in-memory cache.

> **Persistence layer** is a storage we use to read, create, expire, and delete idempotency records.

Reviewer: ```suggestion
**Persistence layer** is a storage we use to create, read, expire, and delete idempotency records.
```
CRUD like? 😄

> **Idempotency record** is the data representation of an idempotent request in its various status. We use it to coordinate  whether **(a)** a request is idempotent, **(b)** it's not expired, **(c)** JSON response to return, and more.

Reviewer: ```suggestion
**Idempotency record** is the data representation of an idempotent request saved in the persistent layer and in its various status. We use it to coordinate  whether **(a)** a request is idempotent, **(b)** it's not expired, **(c)** JSON response to return, and more.
```

| **`dynamodb:PutItem`**{: .copyMe}    | New idempotent records, replace expired idempotent records               |
| **`dynamodb:UpdateItem`**{: .copyMe} | Complete idempotency transaction, and/or update idempotent records state |
| **`dynamodb:DeleteItem`**{: .copyMe} | Delete idempotent records for unsuccessful idempotency transactions      |

**First time setting it up?**

We provide Infrastrucure as Code examples with [AWS Serverless Application Model (SAM)](#aws-serverless-application-model-sam-example), [AWS Cloud Development Kit (CDK)](#aws-cloud-development-kit-cdk), and [Terraform](#terraform) with the required permissions.

### Required resources

To start, you'll need:

<!-- markdownlint-disable MD030 -->

<div class="grid cards" markdown>
*   :octicons-database-16:{ .lg .middle } __Persistent storage__

    ---

    [Amazon DynamoDB](#dynamodb-table) or [Redis](#redis-cluster)

*   :simple-awslambda:{ .lg .middle } **AWS Lambda function**

    ---

    With permissions to use your persistent storage

</div>

<!-- markdownlint-enable MD030 -->

#### DynamoDB table

> !!! tip "You can share a single state table for all functions"

Reviewer: ```suggestion
!!! tip "You can share a single state table for all Lambda functions"
```

Author: by `functions` here, we're referring to both Lambda handler and any Python synchronous function - unless you think it's wrong and we should rethink the messaging altogether

Unless you're looking to use an [existing table or customize each attribute](#dynamodbpersistencelayer), you only need the following:

| Configuration      | Value        | Notes                                                                                    |
| ------------------ | ------------ | ---------------------------------------------------------------------------------------- |

> | Partition key      | `id`         | Format: <br> `{lambda_fn_name}.{module_name}.{fn_qualified_name}#{idempotency_key_hash}` |

Reviewer: Imagine I'm kind of new to the concept of DynamoDB, where do I configure this format in DynamoDB?
It's better to add something like: it will be saved in the format...

If you do not want to accept this suggestion, please correct the format. The correct one is: `{module_name}.{fn_qualified_name}#{idempotency_key_hash}`. Reference: https://github.com/aws-powertools/powertools-lambda-python/blob/develop/aws_lambda_powertools/utilities/idempotency/base.py#L94

Author: will review later, I don't know how to solve it.

Author: crux: we use this for any persistent storage so we need to call it out, but not here.

note to self: test whether we include the Lambda function name

Author: [found the fn name](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/aws_lambda_powertools/utilities/idempotency/persistence/base.py#L116) (more reason why we need to refactor to split responsibilities.. that's supposed to be Idempotency Handler not persistence storage responsibility)

<img width="1520" alt="image" src="https://github.com/user-attachments/assets/91cacc7e-f846-4b9a-a92b-b0d5108c6606">

> | TTL attribute name | `expiration` | Using AWS Console? this is configurable after table creation                             |

Reviewer: ```suggestion
| TTL attribute name | `expiration` | Using AWS Console? This is configurable after table creation                             |
```

    --8<-- "examples/idempotency/templates/terraform.tf"
    ```

##### Limitations

* **DynamoDB restricts [item sizes to 400KB](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#limits-items){target="_blank"}**. This means that if your annotated function's response must be smaller than 400KB, otherwise your function will fail. Consider [Redis](#redis-cluster) as an alternative.

* **Expect 2 WCU per non-idempotent call**. During the first invocation, we use `PutItem` for locking and `UpdateItem` for completion. Consider reviewing [DynamoDB pricing documentation](https://aws.amazon.com/dynamodb/pricing/){target="_blank"} to estimate cost.

* **Old boto3 versions can increase costs**. For cost optimization, we use a conditional `PutItem` to always lock a new idempotency record. If locking fails, it means we already have an idempotency record saving us an additional `GetItem` call. However, this is only supported in boto3 `1.26.194` and higher _([June 30th 2023](https://aws.amazon.com/about-aws/whats-new/2023/06/amazon-dynamodb-cost-failed-conditional-writes/){target="_blank"})_.

> #### Redis cluster

Reviewer: I don't think we should call this section `Redis cluster`. Redis connection/service and can be a standalone instance or a cluster. Why not `Redis Service`?

If you accept this suggestion, please change the link reference in the preview changes.

Author: Good idea! Changed to `Redis database` to follow `DynamoDB table`

We recommend you start with a Redis compatible management services such as [Amazon ElastiCache for Redis](https://aws.amazon.com/elasticache/redis/){target="_blank"} or [Amazon MemoryDB for Redis](https://aws.amazon.com/memorydb/){target="_blank"}.

> In both services and self-hosting Redis, you'll need to configure [VPC access](https://docs.aws.amazon.com/lambda/latest/dg/configuration-vpc.html){target="_blank"} to your AWS Lambda.

Reviewer: We need to review this. If I have a self-hosted solution or any other Redis SaaS solution, I can connect them using the public IP provided by Lambda, for example.

Author: Good catch, removed `self-hosting Redis` to remove ambiguity.

<!-- markdownlint-enable MD013 -->

### Idempotent decorator

For simple use cases, you can use the `idempotent` decorator on your Lambda handler function.

It will treat the entire event as an idempotency key. That is, the same event will return the previously stored result within a [configurable time window](#adjusting-expiration-window) _(1 hour, by default)_.

=== "Idempotent decorator"

    !!! tip "You can also choose [one or more fields](#choosing-a-payload-subset) as an idempotency key."

    ```python title="getting_started_with_idempotency.py" hl_lines="5-8 12 25"
    --8<-- "examples/idempotency/src/getting_started_with_idempotency.py"
    ```

=== "Sample event"

    ```json title="getting_started_with_idempotency_payload.json"
    --8<-- "examples/idempotency/src/getting_started_with_idempotency_payload.json"
    ```

### Idempotent_function decorator

For full flexibility, you can use the `idempotent_function` decorator for any synchronous Python function.

When using this decorator, you **must** call your decorated function using keyword arguments.

> You can use `data_keyword_argument` to tell us the argument to extract an idempotency key.  We support JSON serializable data, [Dataclasses](https://docs.python.org/3.12/library/dataclasses.html){target="_blank" rel="nofollow"}, Pydantic Models, and [Event Source Data Classes](./data_classes.md){target="_blank"}

Reviewer: ```suggestion
You can use `data_keyword_argument` to tell us the argument to extract an idempotency key.  We support JSON serializable data, [Dataclasses](https://docs.python.org/3.12/library/dataclasses.html){target="_blank" rel="nofollow"}, Pydantic Models, and [Event Source Data Classes](./data_classes.md){target="_blank"}.
```

Reviewer: We support to serialize other type of data using CustomDictSerializer, no?

Author: not here, because it's an input (`fn(order: Order)`). Which is a **GREAT** catch because we're missing that functionality today -- similar to Event Handler we should allow customers to bring a custom serializer for input in case they want to bring in a binary of sorts and convert it to JSON before storing in the persistent storage.

`CustomDictSerializer` is for the **output**, which we document under `Any type`

<img width="1160" alt="image" src="https://github.com/user-attachments/assets/860abec9-1365-4274-bf7f-eaa6203c43ae">

The output serializer supports any JSON serializable data, **Python Dataclasses** and **Pydantic Models**.

> !!! info "When using the `output_serializer` parameter, the data will continue to be stored in DynamoDB as a JSON string."

Reviewer: ```suggestion
!!! info "When using the `output_serializer` parameter, the data will continue to be stored in your persistent storage as a JSON string."
```

        ```

=== "Any type"

    Use `CustomDictSerializer` to have full control over the serialization process for any type. It expects two functions:

    * **to_dict**. Function to convert any type to a JSON serializable dictionary before it saves into the persistent storage.
    * **from_dict**. Function to convert from a dictionary retrieved from persistent storage and serialize in its original form.

    ```python hl_lines="9 34 38 42 52 54 64"
    --8<-- "examples/idempotency/src/working_with_idempotent_function_custom_output_serializer.py"
    ```

    1. This function does the following <br><br>**1**. Receives the return from `process_order`<br> **2**. Converts to dictionary before it can be saved into the persistent storage.
    2. This function does the following <br><br>**1**. Receives the dictionary saved into the persistent storage <br>**1** Serializes to `OrderOutput` before `@idempotent` returns back to the caller.
    3. This serializer receives both functions so it knows who to call when to serialize to and from dictionary.

### Using in-memory cache

!!! note "In-memory cache is local to each Lambda execution environment."

> You can enable caching with the `use_local_cache` parameter in `IdempotencyConfig`. When enabled, you can adjust cache capacity _(256)_ with `local_cache_max_items`.

Reviewer: 1024?

<img width="745" alt="image" src="https://github.com/aws-powertools/powertools-lambda-python/assets/4295173/1b4fdd28-823e-4113-a556-81cadb4104b3">

Author: `LRUDict` implementation is 1024, but we reduce in Idempotency to default to 256 to reduce memory footprint.

https://github.com/aws-powertools/powertools-lambda-python/blob/c77869a89c1ebb87718b78b2027a2f0978512142/aws_lambda_powertools/utilities/idempotency/persistence/base.py#L84

    --8<-- "examples/idempotency/src/working_with_payload_subset_payload.json"
    ```

### Adjusting expiration window

> !!! note "We expire idempotency records after **an hour** (3600 seconds). After that, a transaction with the same payload [will not be considered idempotent](#expired-idempotency-records)."

Reviewer: ```suggestion
!!! note "By default, we expire idempotency records after **an hour** (3600 seconds). After that, a transaction with the same payload [will not be considered idempotent](#expired-idempotency-records)."
```

#### RedisPersistenceLayer

!!! info "We recommend Redis version 7 or higher for optimal performance."

> For simple setups, initialize `RedisCachePersistenceLayer` with your cluster endpoint and port to connect.

Reviewer: ```suggestion
For simple setups, initialize `RedisCachePersistenceLayer` with your Redis endpoint and port to connect.
```
